hls-exactprint-utils/src/Ide/TreeTransform.hs

```diff
@@ -0,0 +1,334 @@
+    Graft a <> Graft b = Graft $ \dflags -> a dflags >=> b dflags
+
+instance Monad m => Monoid (Graft m a) where
+    mempty = Graft $ const pure
+
+------------------------------------------------------------------------------
+
+-- | Convert a 'Graft' into a 'WorkspaceEdit'.
+transform ::
+    DynFlags ->
+    ClientCapabilities ->
+    Uri ->
+    Graft (Either String) ParsedSource ->
+    Annotated ParsedSource ->
+    Either String WorkspaceEdit
+transform dflags ccs uri f a = do
+    let src = printA a
+    a' <- transformA a $ runGraft f dflags
+    let res = printA a'
+    pure $ diffText ccs (uri, T.pack src) (T.pack res) IncludeDeletions
+
+------------------------------------------------------------------------------
+
+-- | Convert a 'Graft' into a 'WorkspaceEdit'.
+transformM ::
+    Monad m =>
+    DynFlags ->
+    ClientCapabilities ->
+    Uri ->
+    Graft (ExceptStringT m) ParsedSource ->
+    Annotated ParsedSource ->
+    m (Either String WorkspaceEdit)
+transformM dflags ccs uri f a = runExceptT $
+    runExceptString $ do
+        let src = printA a
+        a' <- transformA a $ runGraft f dflags
+        let res = printA a'
+        pure $ diffText ccs (uri, T.pack src) (T.pack res) IncludeDeletions
+
+------------------------------------------------------------------------------
+
+{- | Construct a 'Graft', replacing the node at the given 'SrcSpan' with the
+ given 'LHSExpr'. The node at that position must already be a 'LHsExpr', or
+ this is a no-op.
+-}
+graft ::
+    forall ast a.
+    (Data a, ASTElement ast) =>
+    SrcSpan ->
+    Located ast ->
+    Graft (Either String) a
+graft dst val = Graft $ \dflags a -> do
+    (anns, val') <- annotate dflags $ maybeParensAST val
+    modifyAnnsT $ mappend anns
+    pure $
+        everywhere'
+            ( mkT $
+                \case
+                    (L src _ :: Located ast) | src == dst -> val'
+                    l -> l
+            )
+            a
+
+------------------------------------------------------------------------------
+
+graftWithM ::
+    forall ast m a.
+    (Fail.MonadFail m, Data a, ASTElement ast) =>
+    SrcSpan ->
+    (Located ast -> TransformT m (Maybe (Located ast))) ->
+    Graft m a
+graftWithM dst trans = Graft $ \dflags a -> do
+    everywhereM'
+        ( mkM $
+            \case
+                val@(L src _ :: Located ast)
+                    | src == dst -> do
+                        mval <- trans val
+                        case mval of
+                            Just val' -> do
+                                (anns, val'') <-
+                                    hoistTransform (either Fail.fail pure) $
+                                        annotate dflags $ maybeParensAST val'
+                                modifyAnnsT $ mappend anns
+                                pure val''
+                            Nothing -> pure val
+                l -> pure l
+        )
+        a
+
+graftWithSmallestM ::
```

Comment:
@konn can you walk me through what this function is supposed to do? My understanding based on the name is that it should only run the transformation on the smallest `Located ast` that contains the span, but I don't see how that could be the case based on the implementation.

Reply:
> can you walk me through what this function is supposed to do?

Sure!

> My understanding based on the name is that it should only run the transformation on the smallest Located ast that contains the span

That's what I originally intended, but not used at last. Here, I used `everywhereM'`, defined [Line 266](https://github.com/haskell/haskell-language-server/pull/759/files/6341b396186a7411fc860e654e63f8559611be7e#diff-a278397f41e59b2252d3f37bbb1795b8758f77305fe719c1bd83de3aa24f5711R266). which is a bottom-up variant of `everywhereM` from `syb`.
Since it is bottom-up, I once expected it rewrites the smallest AST containing the given span.
But now, I think this was mis-implemented: it MUST use bottom-up variant of `somewhere` instead of `everywhereM'`, otherwise it will transform ALL the AST element containing the span.

Reply:
Mind taking a look at my changes in #1543?

Reply:
OK, I'll take a look in this evening (in JST)!